src/table/boot.rs

```diff
@@ -90,6 +90,45 @@ impl BootServices {
         (self.free_pages)(addr as u64, count).into()
     }
 
+    /// Retrieves the current memory map
+    ///
+    /// `buffer` is a location for retrieving the raw memory map from UEFI. This needs to be passed
+    /// in to work around the fact that we have no way to allocate memory and rust does not support variable
+    /// sized stack allocations. We may not use allocate_pages/free_pages as this would change the memory map
+    /// making such an implemenetation unsuitable for supporting usage of ExitBootServices, which requires
+    /// MemoryMap be called and then ExitBootServices be called with the memory map not changing in between.
+    pub fn get_memory_map(&self, output: &mut [MemoryDescriptor], buffer: &mut [u8]) -> Result<(usize, MemoryMapKey)> {
+        // These will get set in the call to memory_map, but we assign default values to
+        // prevent rustc from complaining
+        let mut map_key = MemoryMapKey(0);
+        let mut entry_size = 0;
+        let mut entry_version = 0;
+        // create a subslice of the buffer that is the minimal alignment of a MemoryDescriptor
+        let offset = buffer.as_ptr().align_offset(mem::align_of::<MemoryDescriptor>());
+        if offset > mem::size_of_val(buffer) {
+            return Err(Status::BufferTooSmall);
+        }
+        let aligned_buffer = &mut buffer[offset..];
+        // Attempt to get the result
+        let mut map_size = mem::size_of_val(aligned_buffer);
+        let result = (self.memory_map)(&mut map_size, aligned_buffer.as_mut_ptr() as usize, &mut map_key, &mut entry_size, &mut entry_version);
```

Comment:
We could replace the whole result checking with `(self.memory_map)(...).into()?`. This will convert the Status code into a result and then return early if it's bad.